### PR TITLE
Use a more compact format for debug info

### PR DESCRIPTION
**Pull request by stedolan**
~~This PR generates debug info for each allocation point as well as for each call point, something that's required for generating good backtraces with statmemprof.~~

This PR optimises the size of debug info, as part of #8805 

## Debug info optimisations

The first is to use a more efficient format for inlined frames: instead of a linked-list of pairs of (debuginfo, next), the debuginfo lists for inlined frames are now represented as a contiguous list, using a spare bit in the format to indicate the end of the list. In the common case of no inlined frames, this halves the size of the debug info: instead of 8 bytes of debug info followed by 8 bytes of next pointer (all zeros), there are just 8 bytes of debug info.

The space saving from this optimisation nearly cancels out the loss above: with this change, ocamlopt.opt becomes 0.1% bigger.

## Frame table optimisations

However, the size of the debug info is less important than the size of the frame tables: the frame tables are accessed on a hot path in the GC where cache misses matter, and the debug info is not. Even with the more efficient debug info format, the frame tables get bigger: currently, only 64% of `ocamlopt.opt`'s frame tables have debug info attached. Since the debug info is an 8-byte pointer (on 64-bit machines), this means that trunk currently spends an average of 5 bytes per frame table on debug info.

So, this patch contains a second optimisation, which is to encode the debug info pointers in the frame table as a 32-bit relative offset instead of a full pointer. This means that this patch uses 4 bytes per frame table for debug info, which is a 20% saving over trunk despite adding debug info to all frames.

--- 

With both of these changes, `ocamlopt.opt` actually gets 1.5% *smaller*, despite the extra debug info.

(There remain two situations where it gets bigger though: when compiling without -g, some space is still used in the frame tables for null pointers, and on 32-bit machines relative pointers to debug info don't save any space, so the frame tables get bigger there. If these issues are considered blockers, I can have another go at the frame table format, I think there's more space to be squeezed out there if needed)

### CODE REVIEW

File: runtime/backtrace_nat.c
Comment (xavierleroy) [inline runtime/backtrace_nat.c]: The 24-bit `n` field is now called `f` :-)

**Review by xavierleroy — COMMENTED** [review]
Generally speaking, I'm OK with adding debug info to allocation points, and not too concerned about the increase in frame table size it causes.  I'm obsessed with code size because code cache misses are awfully expensive on modern processors.  But data cache misses are cheaper, the frame table is data, and is not accessed that often (only during GCs and backtrace recording).

Am I correct that only amd64 is done so far and the other emitters need updating?

The change in frame table encoding could have been a separate PR.  At any rate, @let-def should review it, as the author of the original frame table support for inlined functions.  

The one thing I'd like to see is a comment somewhere that explains the representation of frames for inlined functions.

File: runtime/backtrace_nat.c
Comment (stedolan) [reply runtime/backtrace_nat.c]: Thanks! Fixed.

**Comment by stedolan** [comment]
Yes, the extra debuginfo is only enabled for amd64 here. (The frame table format is shared between all architectures, though, so there's only one copy of the encoding/decoding routines).

I've added a comment to emitaux.ml about inlined functions.

**Comment by stedolan** [comment]
Superseded by #8805

File: runtime/roots_nat.c
Comment (jhjourdan) [inline runtime/roots_nat.c:32]: Is this useful?

File: runtime/roots_nat.c
Comment (jhjourdan) [inline runtime/roots_nat.c:83]: Any reason for testing `>= 4096` instead of `!= 0`? I guess the reason this is true is because of the size of pages, but I do not see why asserting this is interesting here.

File: runtime/roots_nat.c
Comment (jhjourdan) [inline runtime/roots_nat.c:82]: Perhaps you could add a (temporary) comment saying what `num_alloc` is designed for.

File: asmcomp/amd64/emit.mlp
Comment (jhjourdan) [inline asmcomp/amd64/emit.mlp:1131]: Where is this label used? Is there any place in the runtime where this piece of information is read?

**Review by jhjourdan — CHANGES_REQUESTED** [review]
This changes look OK for me, apart from a few comments I added in this review. In particular, I do not understand the purpose of the extra `D.size` emitted in `amd64/emit.mlp`.

I have one concern, though: we are more and more using 32-bits offsets instead of pointers for saving various data structures in the executable. This is fine as long as the executable does not get too big, but what will happen if these offset overflow, in case of a giant executable (I heard that JS folks are sometimes producing very large executables)? Will `ocamlopt` or the assembler complain about this, or will the compilation chain simply generate an invalid executable?

And well, of course, some work needs to be done for porting/testing the other architectures before merging.

File: runtime/roots_nat.c
Comment (stedolan) [reply runtime/roots_nat.c:83]: It is indeed true because of the size of pages. During development, this caught several bugs that the `!= 0` check missed, where a misalignment of frametable data caused the retaddr to contain a small integer (e.g. frame size) instead of the return address.

File: asmcomp/amd64/emit.mlp
Comment (stedolan) [reply asmcomp/amd64/emit.mlp:1131]: This size information is not read by the runtime.

This exports an ELF size, so that standard ELF binary tools (e.g. `objdump`, `readelf`, etc.) can be used to measure the size of a frame table. (Before this change, such tools show the frametable as having size 0). This information is already exported for OCaml functions, and for assembly stubs since #8708.

**Comment by stedolan** [comment]
> I have one concern, though: we are more and more using 32-bits offsets instead of pointers for saving various data structures in the executable. This is fine as long as the executable does not get too big, but what will happen if these offset overflow, in case of a giant executable (I heard that JS folks are sometimes producing very large executables)? Will ocamlopt or the assembler complain about this, or will the compilation chain simply generate an invalid executable?

There are, as far as I know, only two places where 32-bit offsets are used instead of pointers:

  - In debug info, to point from the location info to the filename (already present)
  - In frametables, to point to the debug info (introduced by this PR)

The important point is that both of these pointers point within a compilation unit, and never across modules. So, in order to overflow the offset, we would need not a 4GB executable but a 4GB *module*. In fact, since the debug info and frame tables are stored after the code, we'd need a module with a 4GB frame table.

It's true that JS does produce large executables, with the largest I'm aware of being on the order of 1GB. However, these are composed of many smaller modules, some up to megabytes but I found none with more than ~200K of frame table.

The largest single module that I found while testing this design is not a JS module, but one produced by Ahrefs and posted here in a bug report (#7452). This module involves a large amount of generated code (approx. 14 MB of source code), and compiles (eventually!) to a 21MB object file, which has a ~3MB frame table (including debug info). This would still need to be 1000x bigger to overflow 32 bits.

(The existence of that module is why a more aggressive packing design using only 24-bit debuginfo pointers was abandoned. 3MB is uncomfortably close to a 16MB limit).

File: runtime/roots_nat.c
Comment (stedolan) [reply runtime/roots_nat.c:82]: Done. (Added below, where `num_allocs` is assigned)

File: runtime/roots_nat.c
Comment (stedolan) [reply runtime/roots_nat.c:32]: No!

File: runtime/roots_nat.c
Comment (stedolan) [reply runtime/roots_nat.c:82]: Actually, I've just deleted that variable for now.

**Comment by jhjourdan** [comment]
> The important point is that both of these pointers point within a compilation unit, and never across modules. So, in order to overflow the offset, we would need not a 4GB executable but a 4GB module. In fact, since the debug info and frame tables are stored after the code, we'd need a module with a 4GB frame table.

Right. So the risk is indeed very limited. Perhaps we could add assertions in the emitter just to make sure this never happens.

**Comment by stedolan** [comment]
> Right. So the risk is indeed very limited. Perhaps we could add assertions in the emitter just to make sure this never happens.

We can't add an assertion to the emitter, because the emitter doesn't compute these offsets. Instead, it outputs an assembler directive to give a 32-bit offset to a label. I think the assembler will give an error if the offset doesn't fit, assuming the ocaml compiler can produce and the assembler can consume a >4GB file.

**Comment by jhjourdan** [comment]
Now that #8805 is merged, could we close this PR?

**Comment by stedolan** [comment]
Yes, thanks for the reminder.

**Pull request closed without merging**
